lua/ttt2/libraries/migrations.lua

```diff
@@ -0,0 +1,122 @@
+
+---
+-- Runs all missing forward gamemode migrations.
+-- @return boolean `true` if the desired version was successfully migrated and `false` in case of an error.
+-- @realm shared
+function migrations.Apply()
+	if not migrations.isLoaded then
+		migrations.Load()
+	end
+
+	local migrationSuccess = true
+	local errorMessage = ""
+
+	for fileName, migration in SortedPairs(migrations.migrations) do
+		if not IsValid(migration) then continue end
+
+		local fileInfo = migrations.orm:Find(fileName)
+
+		-- As ORM currently gives out strings check it
+		if fileInfo and (fileInfo.wasSuccess == true or fileInfo.wasSuccess == "true") then continue end
+
+		MsgN("[TTT2] Migrating: ", fileName)
+
+		migrationSuccess, errorMessage = pcall(migration.Upgrade, migration)
+
+		local executedAt = os.date("%Y/%m/%d - %H:%M:%S", os.time())
+
+		if fileInfo then
+			fileInfo.executedAt = executedAt
+			fileInfo.statusMessage = migrationSuccess and "Success" or errorMessage
+			fileInfo:Save()
+		else
+			migrations.orm:New({
+				name = fileName,
+				executedAt = executedAt,
+				statusMessage = migrationSuccess and "Success" or errorMessage,
+			}):Save()
+		end
+
+		if not migrationSuccess then
+			ErrorNoHalt("[TTT2] Migration failed. Error:\n" .. tostring(errorMessage) .. "\n")
+			break;
+		else
+			MsgN("[TTT2] Successfully migrated: ", fileName)
+		end
+	end
+
+	return migrationSuccess
+end
+
+---
+-- Checks if this class should inherit from the baseClass
+-- In this case only looks up, if the class is the base
+-- @param table class The class that inherits
+-- @param table base The class to inherit from
+-- @return boolean If the class is the base
+local function ShouldInherit(class, baseClass)
+	return class.base ~= class.type
+end
+
+---
+-- Run on Initialization of a class after building it
+-- @param table class The Class that got initialized
+-- @param string path The path the file was found in
+-- @param string name The name of the class
+local function OnInitialization(class, path, name)
+	class.type = name
+	class.base = class.base or "migration_base"
+
+	MsgN("Added TTT2 Migration file: ", path, name)
+end
+
+---
+-- Run after inheriting to set as final baseclass
+-- @param table class The Class that got initialized
+-- @param string path The path the file was found in
+-- @param string name The name of the class
+local function PostInherit(class, path, name)
+	baseclass.Set(name, class)
+end
+
+---
+-- Loads migration-files and starts inheriting from base files
+-- @realm shared
+function migrations.Load()
+	MsgN("[TTT2] Loading base Migrations for inheritance ...")
+	local baseClasses = classbuilder.BuildFromFolder(migrations.baseFolderPath, SHARED_FILE, "MIGRATION", OnInitialization, true, ShouldInherit, {}, PostInherit)
+
+	MsgN("[TTT2] Loading Migrations ...")
+	migrations.migrations = classbuilder.BuildFromFolder(migrations.folderPath, SHARED_FILE, "MIGRATION", OnInitialization, true, ShouldInherit, baseClasses, PostInherit)
```

Comment:
The class builder is also rather overkill. I would not like to have the inheritance included for migrations. Just make the migrations return the up and down function and load the files.

Reply:
I actually like it way more. With the earlier command setup I wanted to give a way for default behaviours.
With this inheritance instead we can make sure to program a few base migrations and reuse them over and over again. Makes it way safer, then copying code over and over again, IMHO.

Reply:
Migrations are really nearly never connected to each other. You will never need this as far as i can tell from my experience

Reply:
i don't think it hurts. if i wanted to implement a generic convar migration or some other situation it would save some juggling.

Reply:
Yes, thats what I thought too.
We wanted:
- Create databases here
- Rename some databases
- Maybe change savingKeys
- Change convarnames
- Change convardefaults

I cant really think of much more we ever did, having those few cases standardized was my goal with this PR. We can implement them along the way and can be sure they work.

Reply:
I strongly disagree. Implementing stuff just because it does not hurt is not our ambition here, we would rather have something simple and clear to solve the problem.
I think you both are only thinking theoretically about this. You wont need any of this through inheritance. This concept is really nothing that brings benefits here. We do not need any hierarchical structure to provide reusable tools to do these common changes but rather utilities. 
We should focus on making this as accessible and easy as possible. It is way easier to just write a migration file that returns a single function ("upgrade" or "migrate") or maybe two if we want to keep both directions.

Composition based on utilities is much more flexible for such a scenario, where object oriented design is rather inflexible and does not allow to be combined as much. Especially since we do not need any of the structural benefits that come with such a design. So we only get the downsides.

Reply:
That I can agree with. We definitely need clear utilities. I thought that classes can give us exactly that. But in our case migrations are limited to maybe 2? areas. We dont need to handle files or any other Apis.

For me it's important that stuff that gets used more than once is never rewritten.
So I strongly disagree to have any viable code in the migration files that get dirctly applied itself, they have to be inherited or given by utilities. And IMO should therefore also directly allow for downgrade, even if we never properly implement it.

And even if we dont want to inherit, using the classbuilder is a plus IMO.
The IsValid check can be surely done by the migration.apply function, but using the Migration globals seems useful.

Reply:
I can't seem to get across my points to you, but i still want to say that this is just making things unnecessarily complex and migrations do not share code. When looking at bigger Laravel projects for example i know that they would not benefit from inheritance.
Also the entire base migration has no use. It does not provide any functionality that makes it reasonable to introduce this complexity. Having to define a base class for migrations is just not needed. We just get more dead code for some potentially highly theoretical benefit, that we do not even know a use case of yet. This is rather over-engineering the problem to be honest.

What are your opinions @Histalek or @TimGoll ?